cpp/modmesh/universe/bezier.hpp

```diff
@@ -0,0 +1,216 @@
+    void set_y(value_type v) { y() = v; }
+
+    value_type z() const { return m_coord[2]; }
+    value_type & z() { return m_coord[2]; }
+    void set_z(value_type v) { z() = v; }
+
+    T operator[](size_t i) const { return m_coord[i]; }
+    T & operator[](size_t i) { return m_coord[i]; }
+
+    T at(size_t i) const
+    {
+        check_size(i, 3);
+        return m_coord[i];
+    }
+    T & at(size_t i)
+    {
+        check_size(i, 3);
+        return m_coord[i];
+    }
+
+    size_t size() const { return 3; }
+
+    void fill(T v) { m_coord[0] = m_coord[1] = m_coord[2] = v; }
+
+private:
+
+    void check_size(size_t i, size_t s) const
+    {
+        if (i >= s)
+        {
+            throw std::out_of_range(Formatter() << "Vector3d: i " << i << " >= size " << s);
+        }
+    }
+
+    T m_coord[3];
+
+}; /* end class Vector3d */
+
+using Vector3dFp32 = Vector3d<float>;
+using Vector3dFp64 = Vector3d<double>;
+
+/**
+ * Bezier curve in three-dimensional space.
+ *
+ * @tparam T floating-point type
+ */
+template <typename T>
+class Bezier3d
+    : public NumberBase<int32_t, T>
+{
+
+public:
+
+    using vector_type = Vector3d<T>;
+
+    explicit Bezier3d(std::vector<vector_type> const & controls)
+        : m_controls(controls)
+    {
+    }
+
+    Bezier3d() = default;
+    Bezier3d(Bezier3d const &) = default;
+    Bezier3d(Bezier3d &&) = default;
+    Bezier3d & operator=(Bezier3d const &) = default;
+    Bezier3d & operator=(Bezier3d &&) = default;
+    ~Bezier3d() = default;
+
+    size_t size() const { return ncontrol(); }
+    vector_type const & operator[](size_t i) const { return control(i); }
+    vector_type & operator[](size_t i) { return control(i); }
+    vector_type const & at(size_t i) const { return control_at(i); }
+    vector_type & at(size_t i) { return control_at(i); }
+
+    size_t ncontrol() const { return m_controls.size(); }
+    vector_type const & control(size_t i) const { return m_controls[i]; }
+    vector_type & control(size_t i) { return m_controls[i]; }
+    vector_type const & control_at(size_t i) const
+    {
+        check_size(i, m_controls.size(), "control");
+        return m_controls[i];
+    }
+    vector_type & control_at(size_t i)
+    {
+        check_size(i, m_controls.size(), "control");
+        return m_controls[i];
+    }
+
+    size_t nlocus() const { return m_loci.size(); }
+    vector_type const & locus(size_t i) const
+    {
+        check_size(i, m_loci.size(), "locus");
+        return m_loci[i];
+    }
+    vector_type & locus(size_t i)
+    {
+        check_size(i, m_loci.size(), "locus");
+        return m_loci[i];
+    }
+
+    void sample(size_t nlocus);
+
+private:
+
+    void check_size(size_t i, size_t s, char const * msg) const
+    {
+        if (i >= s)
+        {
+            throw std::out_of_range(Formatter() << "Bezier3d: (" << msg << ") i " << i << " >= size " << s);
+        }
+    }
+
+    std::vector<vector_type> m_controls;
```

Comment:
The prototype uses a dynamic container.  The overhead may probably be unavoidable.  But it needs to be improved in the future.